main.cc

```diff
@@ -1148,6 +1148,11 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             auth_prep_cache_config.refresh = std::chrono::milliseconds(cfg->permissions_update_interval_in_ms());
 
             qp.start(std::ref(proxy), std::ref(forward_service), std::move(local_data_dict), std::ref(mm_notifier), std::ref(mm), qp_mcfg, std::ref(cql_config), std::move(auth_prep_cache_config), std::ref(group0_client)).get();
+
+            ss.invoke_on_all([&] (service::storage_service& ss) {
+                ss.set_query_processor(qp.local());
```

Comment:
ss.start() is called before qp.start(), so qp is not ready when we invoke the storage_service constructor. From the point of view of the implementation of storage_service, taking something by the constructor means that this thing should be ready to use. It's not just about passing the pointer. 

We used to have those pointers globally available to everyone and it created a mess because you couldn't know at usage site if is actually ready to use. 

The set_query_processor() thing is ugly, but at least you know in the implementation that you cannot trust it and be extra careful.

Reply:
> From the point of view of the implementation of storage_service, taking something by the constructor means that this thing should be ready to use. It's not just about passing the pointer.

Agree, I just missed that `storage_service` was initialized much earlier.